### ISSUE TITLE
[FR] Add Preview For JupyterNotebook files .ipynb

### ISSUE THREAD

**Issue by AdnenKhiari** [mlflow/mlflow#9650]
### Willingness to contribute

Yes. I would be willing to contribute this feature with guidance from the MLflow community.

### Proposal Summary

**What is the use case for this feature ?**
Be Able to preview code from the ui server without having to download the file to check the code
**Why is this use case valuable to support for MLflow users in general?**
Better efficiency 
**Why is this use case valuable to support for your project(s) or organization?**
As an mlflow user , it would help me to be able to visualize my notebooks in the ui
**Why is it currently difficult to achieve this use case? (please be as specific as possible about why related MLflow features and components are insufficient)**
Currently we can visualize image, text, html, pdf, json files but not jupyter notebooks 

### Motivation

> #### What is the use case for this feature?

> #### Why is this use case valuable to support for MLflow users in general?

> #### Why is this use case valuable to support for your project(s) or organization?

> #### Why is it currently difficult to achieve this use case?

### What component(s) does this bug affect?

- [ ] `area/artifacts`: Artifact stores and artifact logging
- [ ] `area/build`: Build and test infrastructure for MLflow
- [ ] `area/docs`: MLflow documentation pages
- [ ] `area/examples`: Example code
- [ ] `area/gateway`: AI Gateway service, Gateway client APIs, third-party Gateway integrations
- [ ] `area/model-registry`: Model Registry service, APIs, and the fluent client calls for Model Registry
- [ ] `area/models`: MLmodel format, model serialization/deserialization, flavors
- [ ] `area/recipes`: Recipes, Recipe APIs, Recipe configs, Recipe Templates
- [ ] `area/projects`: MLproject format, project running backends
- [ ] `area/scoring`: MLflow Model server, model deployment tools, Spark UDFs
- [ ] `area/server-infra`: MLflow Tracking server backend
- [ ] `area/tracking`: Tracking Service, tracking client APIs, autologging

### What interface(s) does this bug affect?

- [ ] `area/uiux`: Front-end, user experience, plotting, JavaScript, JavaScript dev server
- [ ] `area/docker`: Docker use across MLflow's components, such as MLflow Projects and MLflow Models
- [ ] `area/sqlalchemy`: Use of SQLAlchemy in the Tracking Service or Model Registry
- [ ] `area/windows`: Windows support

### What language(s) does this bug affect?

- [ ] `language/r`: R APIs and clients
- [ ] `language/java`: Java APIs and clients
- [ ] `language/new`: Proposals for new client languages

### What integration(s) does this bug affect?

- [ ] `integrations/azure`: Azure and Azure ML integrations
- [ ] `integrations/sagemaker`: SageMaker integrations
- [ ] `integrations/databricks`: Databricks integrations

**Comment by BenWilson2** [comment]
@sunishsheth2009 @harupy what are your thoughts on this? 
Would these render correctly and what are any issues with attempting to get the artifact viewer pane to display such a file?

**Comment by sunishsheth2009** [comment]
Yes this surely seems like a good idea. :) 
We need to find a library that renders this correctly without impacting the performance of rendering it. 

cc @hubertzub-db to get some thoughts as well.

**Comment by hubertzub-db** [comment]
Hey @AdnenKhiari, thanks for filing this!
So there's only one library I've managed to find at the moment: https://www.npmjs.com/package/notebookjs
It seems to be quite old, but is still occasionally contributed to so it might be worth checking it out. Also, like @sunishsheth2009 mentioned, let's confirm that the library's performance is satisfying.
Additional thoughts on this:
- I believe that for safety reasons, the initial implementation should have JS execution disabled until we figure out some sandboxing mechanism
- The wrapper component responsible for rendering this should be lazy loaded `import (() => '...')` so the imported library will end up in separate chunk asset

Happy to answer any other questions if necessary! Also @AdnenKhiari please let us know if you know about any other 3rd party libraries capable of rendering Jupyter files

**Comment by harupy** [comment]
@AdnenKhiari Can you save notebooks as HTML files?

<img width="1621" alt="image" src="https://github.com/mlflow/mlflow/assets/17039389/0087431f-68f4-4927-a8ee-fa3b0c97ec2f">

**Comment by harupy** [comment]
or we can make the backend render a notebook:

<img width="1253" alt="image" src="https://github.com/mlflow/mlflow/assets/17039389/8dc9ef91-143d-43bb-b40a-9d577cf76d5c">

**Comment by AdnenKhiari** [comment]
A simpler approach could be to automatically save a HTML file version each time a jupyter notebook is logged to the artifacts , what are your thoughts ? instead of converting it each time it is in preview

**Comment by harupy** [comment]
@AdnenKhiari What do you think about `mlflow.log_notebook`?

**Comment by AdnenKhiari** [comment]
Yes i think it is a good idea , shall i give it a try and fire a PR ?

**Comment by harupy** [comment]
I'm also curious how mlflow's competitors such as Wandb render notebooks.

**Comment by AdnenKhiari** [comment]
Now that i think of it , if we just log the html version , it's true that preview is now easy , but if someone needs the code he logged , he will not be able to use , as it will be an html file , so we need to store the source file and just use the html for the preview , I suggest to actually save both the ipynb file and an html version with it , so that we can download the code when we need . We can even make changes such that the html file corresponding to the jupyter file becomes hidden in the ui and is used automatically for previewing the notebook , what do you think ? @harupy

**Comment by harupy** [comment]
> Now that i think of it , if we just log the html version , it's true that preview is now easy , but if someone needs the code he logged

This is the reason why I was leaning towards the render-on-ui approach.

**Comment by AdnenKhiari** [comment]
that's true , another alternative following @hubertzub-db comment is to use this library  https://github.com/Joeyonng/react-jupyter-notebook , we can solve everything by just using a component that renders jupyter notebooks and allows for some configuration too
